src/fnft_nsev.c

```diff
@@ -156,220 +177,366 @@ INT fnft_nsev(
+        opts->bound_state_localization = nsev_bsloc_NEWTON;
+        ret_code = fnft_nsev_base(D_effective, q_preprocessed, r_preprocessed, T, M, contspec, XI, K_ptr,
+                bound_states, normconsts_or_residues_reserve, kappa, opts);
+        CHECK_RETCODE(ret_code, leave_fun);
+        
+        // Restore original state of opts
+        opts->bound_state_localization = nsev_bsloc_SUBSAMPLE_AND_REFINE;
+    } else { 
+        ret_code = fnft_nsev_base(D_effective, q_preprocessed, r_preprocessed, T, M, contspec, XI, K_ptr,
+                    bound_states, normconsts_or_residues, kappa, opts);
+        CHECK_RETCODE(ret_code, leave_fun);
+    }
+    
+    if (opts->richardson_extrapolation_flag == 1){
+        // Allocating memory
+        UINT contspec_len = 0;
+        if (contspec != NULL && M > 0){
+            switch (opts->contspec_type) {
+                case nsev_cstype_BOTH:
+                    contspec_len = 3*M;
+                    break;
+                case nsev_cstype_REFLECTION_COEFFICIENT:
+                    contspec_len = M;
+                    break;
+                case nsev_cstype_AB:
+                    contspec_len = 2*M;
+                    break;
+                default:
+                    ret_code = E_INVALID_ARGUMENT(opts->contspec_type);
+                    goto leave_fun;
+            }
+            contspec_sub = malloc(contspec_len * sizeof(COMPLEX));
+            if (contspec_sub == NULL) {
+                ret_code = E_NOMEM;
+                goto leave_fun;
+            }
         }
-        // Norming constants and/or residues)
-        if (normconsts_or_residues != NULL && *K_ptr != 0) {
-
-            ret_code = tf2normconsts_or_residues(D, q, T, *K_ptr,
-                transfer_matrix, deg, bound_states, normconsts_or_residues,
-                opts);
-            CHECK_RETCODE(ret_code, release_mem);
+        UINT discspec_len = 0;
+        if (kappa == +1 && bound_states != NULL && *K_ptr != 0) {
+            K_sub = *K_ptr;
+            bound_states_sub = malloc(K_sub * sizeof(COMPLEX));
+            discspec_len = K_sub;
+            switch (opts->discspec_type) {
+                case nsev_dstype_BOTH:
+                case nsev_dstype_RESIDUES:
+                    discspec_len = 2*K_sub;
+                    break;
+                case nsev_dstype_NORMING_CONSTANTS:
+                    discspec_len = K_sub;
+                    break;
+                default:
+                    ret_code = E_INVALID_ARGUMENT(opts->discspec_type);
+                    goto leave_fun;
+            }
+            normconsts_or_residues_sub = malloc(discspec_len * sizeof(COMPLEX));
+            if (normconsts_or_residues_sub == NULL || bound_states_sub == NULL) {
+                ret_code = E_NOMEM;
+                goto leave_fun;
+            }
+            for (i=0; i<K_sub; i++)
+                bound_states_sub[i] = bound_states[i];
+        }
+        // Preparing q_preprocessed
```

Comment:
Maybe mention that in contrast to before, this q now is subsampled to clarify. Richardson combines the result of `fnft_nsev_base` for this signal with the previous, not subsampled run.